### PR TITLE
Autowiring for all!

### PR DESCRIPTION
**Pull request by escopecz**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [N]
| New feature/enhancement? (use the a.x branch)      | [Y]
| Deprecations?                          | [N]
| BC breaks? (use the c.x branch)        | [Y]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | https://github.com/mautic/developer-documentation-new/pull/89
| Issue(s) addressed                     | Fixes #... 

#### Description:

Symfony can autowire and autoconfigure services for some time. Here is the doc:

https://symfony.com/doc/4.4/service_container.html

Mautic was still continuing the old way where each service must be configured in the config.php file with all its dependencies. It is a bit daunting to keep all the dependencies up to date in the config file. So this PR is making the developers work a bit nicer.

1. New services do not need to be configured in config.php files
2. Old services can drop their parameters and keep the aliases for BC
3. Did you know that commands were autowired since M3? We still configured them in config.php files...

The biggest problem here is with classes that are not services. Like DTO, value objects, events (which is also DTO) or exceptions. We have to exclude those. In `app/config/services.php` there is common configuration of those excluded directories for all bundles including plugins. But each bundle and plugin can define its own excludes in *Bundle/Config/services.php.

#### Next steps

As this PR is big already I do not want to continue the service refactoring and rather start a new PR for each bundle. This PR enables the refactoring.

1. Finish converting all services in config.php to just a list of aliases in services.php bundle by bundle.
2. Refactor all controllers to use DI
3. Set all services as private by default.

#### Steps to test this PR:

I know this is hard to swallow, but this PR is one of those that can break any single feature in the project.

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
4. The container build would fail if there would be a bug in the changed code. So Mautic wouldn't even load.
5. But I was testing with `bin/console debug:autowiring` command to show me that all services are wired correctly. Clear cache in between runs and changes you make.
6. For example I removed service definition for the command `mautic:segments:update` and service it uses to show that it can be done. So you can run that it works.
7. An issue can come in places where we lazy-load services which is in controllers mainly. So clicking through the UI would be a good test.

### CODE REVIEW

**Comment by biozshock** [comment]
Maybe instead of `serviceAliases` we ~~should add a pass~~:
```php
class AliasPass implements CompilerPassInterface
{

    public function process(ContainerBuilder $container): void
    {
        foreach ($container->getDefinitions() as $definitionId => $definition) {
            if ($definition->isAbstract()) {
                continue;
            }

            if ($definition->getClass() === '' || $definition->getClass() === null) {
                continue; // most of these are childDefinitions
            }

            if ($definitionId === $definition->getClass()) {
                continue;
            }

            if ($container->hasAlias($definitionId) || $container->hasAlias($definition->getClass())) {
                continue;
            }

            if (strpos($definitionId, '\\') !== false) {
                continue;
            }
            
            $container->removeDefinition($definitionId);
            $container->setDefinition($definition->getClass(), $definition);

            $container->setAlias($definitionId, new Alias($definition->getClass(), $definition->isPublic()));
        }
    }

}
```

EDIT: The pass may actually remove some of the symfony's services, which is not really good, so best way to mitigate this is to add the `LegacyAutowire` trait with private method above and include it to each of the `*Extension.php`

**Comment by biozshock** [comment]
Also would it be possible to refactor `config.php` files to symfony `services.php` and load classes from there? That will allow to get rid of a huge excludes in `app/config/services.php` and will allow to apply this improvement to plugins? e.g. if the services.php exists the code will be symfony autoloaded and if not then the older config.php will be used.

**Comment by escopecz** [comment]
@biozshock thanks for the suggestions! So you suggest to leave the service configuration as is, maybe remove the `arguments` array and leave the `class` and service name?

The short term plan I have in mind is that we'd enable autowiring to be used for M5 but the service aliases would still exit. In M6 we'd remove the aliases too. Is that something you and others agree with?

Allowing bundles and plugins to configure their own excludes would be great. I'll look into it.

**Comment by escopecz** [comment]
@biozshock I'm playing with a bundle having option to have its own `Config/service.php` to exclude its own files. I'm afraid we'll have to choose whether we want to let the Core bundle configure it for all bundles and plugins or force each bundle to configure its own namespace. It cannot be both. What would you prefer?

**Comment by biozshock** [comment]
@escopecz sure thing - each bundle must configure itself. As only bundle knows what's need to be be configured and how.
As far as i can see the services.php will contain:

1. a default configuration autowire:true, autoconfigure: true, public: false
2. a bunch of aliases to current string service names, if we are getting rid of `config.php` completely.
3. services that can not be autoconfigured automaticelly. But please don't use indexed arguments, but a named ones.

Maybe the first iteration should have only point 1. Because getting rid of config.php in the current state is not really an easy task (tried that locally, but have couple of ideas). Introducing services.php will allow us to autowire at least the core services and default plugins and for plugin authors to use autowiring.

**Comment by escopecz** [comment]
@biozshock makes sense. I found a way how to keep the configuration for bundles (including plugins) that do not need to exclude anything special and let other bundles to configure their own services if needed. This is quick&dirty implementation with showcase on LeadBundle and a plugin:

https://github.com/mautic/mautic/pull/11587/commits/d695f15f128fc558209b8411bde4ac284fe002fd

WDYT?

So the next work would be to move the excluded services in app/config/services.php to their own bundle configurations.

**Comment by escopecz** [comment]
@biozshock I tried the trait you suggested but it creates a loop:
```
  [Symfony\Component\DependencyInjection\Exception\ServiceCircularReferenceException]                                                                       
  Circular reference detected for service "LightSaml\Store\Credential\CompositeCredentialStore", path: "LightSaml\Store\Credential\CompositeCredentialStore -> LightSaml\Store\Credential\CompositeCredentialStore". 
```

**Review by RCheesley — CHANGES_REQUESTED** [review]
Thanks for all the work on this folks!

1. I have found an issue with the builder:

When opening emails or landing pages, the old legacy builder is being loaded but GrapesJS is enabled, so that should be loaded.

2. When creating users the following error occurs:

```
[2022-10-26 13:46:36] mautic.CRITICAL: Uncaught PHP Exception TypeError: "Mautic\CoreBundle\Translation\Translator::trans(): Argument #1 ($id) must be of type string, null given, called in /var/www/html/vendor/symfony/framework-bundle/Templating/Helper/TranslatorHelper.php on line 56" at /var/www/html/app/bundles/CoreBundle/Translation/Translator.php line 41 {"exception":"[object] (TypeError(code: 0): Mautic\\CoreBundle\\Translation\\Translator::trans(): Argument #1 ($id) must be of type string, null given, called in /var/www/html/vendor/symfony/framework-bundle/Templating/Helper/TranslatorHelper.php on line 56 at /var/www/html/app/bundles/CoreBundle/Translation/Translator.php:41)
[stacktrace]
#0 /var/www/html/vendor/symfony/framework-bundle/Templating/Helper/TranslatorHelper.php(56): Mautic\\CoreBundle\\Translation\\Translator->trans(NULL, Array, NULL, NULL)
#1 /var/www/html/vendor/symfony/framework-bundle/Resources/views/Form/attributes.html.php(3): Symfony\\Bundle\\FrameworkBundle\\Templating\\Helper\\TranslatorHelper->trans(NULL, Array, NULL)
#2 /var/www/html/app/bundles/CoreBundle/Templating/Engine/PhpEngine.php(147): require('/var/www/html/v...')
#3 /var/www/html/vendor/symfony/templating/PhpEngine.php(76): Mautic\\CoreBundle\\Templating\\Engine\\PhpEngine->evaluate(NULL, Array)
#4 /var/www/html/app/bundles/CoreBundle/Templating/Engine/PhpEngine.php(119): Symfony\\Component\\Templating\\PhpEngine->render('FrameworkBundle...', Array)
#5 /var/www/html/vendor/symfony/form/Extension/Templating/TemplatingRendererEngine.php(41): Mautic\\CoreBundle\\Templating\\Engine\\PhpEngine->render('FrameworkBundle...', Array)
#6 /var/www/html/vendor/symfony/form/FormRenderer.php(115): Symfony\\Component\\Form\\Extension\\Templating\\TemplatingRendererEngine->renderBlock(Object(Symfony\\Component\\Form\\FormView), 'FrameworkBundle...', 'attributes', Array)
#7 /var/www/html/vendor/symfony/framework-bundle/Templating/Helper/FormHelper.php(211): Symfony\\Component\\Form\\FormRenderer->renderBlock(Object(Symfony\\Component\\Form\\FormView), 'attributes', Array)
#8 /var/www/html/vendor/symfony/framework-bundle/Resources/views/Form/widget_container_attributes.html.php(2): Symfony\\Bundle\\FrameworkBundle\\Templating\\Helper\\FormHelper->block(Object(Symfony\\Component\\Form\\FormView), 'attributes')
#9 /var/www/html/app/bundles/CoreBundle/Templating/Engine/PhpEngine.php(147): require('/var/www/html/v...')
#10 /var/www/html/vendor/symfony/templating/PhpEngine.php(76): Mautic\\CoreBundle\\Templating\\Engine\\PhpEngine->evaluate(NULL, Array)
#11 /var/www/html/app/bundles/CoreBundle/Templating/Engine/PhpEngine.php(119): Symfony\\Component\\Templating\\PhpEngine->render('FrameworkBundle...', Array)
#12 /var/www/html/vendor/symfony/form/Extension/Templating/TemplatingRendererEngine.php(41): Mautic\\CoreBundle\\Templating\\Engine\\PhpEngine->render('FrameworkBundle...', Array)
#13 /var/www/html/vendor/symfony/form/FormRenderer.php(115): Symfony\\Component\\Form\\Extension\\Templating\\TemplatingRendererEngine->renderBlock(Object(Symfony\\Component\\Form\\FormView), 'FrameworkBundle...', 'widget_containe...', Array)
#14 /var/www/html/vendor/symfony/framework-bundle/Templating/Helper/FormHelper.php(211): Symfony\\Component\\Form\\FormRenderer->renderBlock(Object(Symfony\\Component\\Form\\FormView), 'widget_containe...', Array)
#15 /var/www/html/app/bundles/CoreBundle/Views/FormTheme/Custom/form_widget_compound.html.php(1): Symfony\\Bundle\\FrameworkBundle\\Templating\\Helper\\FormHelper->block(Object(Symfony\\Component\\Form\\FormView), 'widget_containe...')
#16 /var/www/html/app/bundles/CoreBundle/Templating/Engine/PhpEngine.php(147): require('/var/www/html/a...')
#17 /var/www/html/vendor/symfony/templating/PhpEngine.php(76): Mautic\\CoreBundle\\Templating\\Engine\\PhpEngine->evaluate(NULL, Array)
#18 /var/www/html/app/bundles/CoreBundle/Templating/Engine/PhpEngine.php(119): Symfony\\Component\\Templating\\PhpEngine->render('MauticCoreBundl...', Array)
#19 /var/www/html/vendor/symfony/form/Extension/Templating/TemplatingRendererEngine.php(41): Mautic\\CoreBundle\\Templating\\Engine\\PhpEngine->render('MauticCoreBundl...', Array)
#20 /var/www/html/vendor/symfony/form/FormRenderer.php(115): Symfony\\Component\\Form\\Extension\\Templating\\TemplatingRendererEngine->renderBlock(Object(Symfony\\Component\\Form\\FormView), 'MauticCoreBundl...', 'form_widget_com...', Array)
#21 /var/www/html/vendor/symfony/framework-bundle/Templating/Helper/FormHelper.php(211): Symfony\\Component\\Form\\FormRenderer->renderBlock(Object(Symfony\\Component\\Form\\FormView), 'form_widget_com...', Array)
#22 /var/www/html/vendor/symfony/framework-bundle/Resources/views/Form/form_widget.html.php(2): Symfony\\Bundle\\FrameworkBundle\\Templating\\Helper\\FormHelper->block(Object(Symfony\\Component\\Form\\FormView), 'form_widget_com...')
#23 /var/www/html/app/bundles/CoreBundle/Templating/Engine/PhpEngine.php(147): require('/var/www/html/v...')
#24 /var/www/html/vendor/symfony/templating/PhpEngine.php(76): Mautic\\CoreBundle\\Templating\\Engine\\PhpEngine->evaluate(NULL, Array)
#25 /var/www/html/app/bundles/CoreBundle/Templating/Engine/PhpEngine.php(119): Symfony\\Component\\Templating\\PhpEngine->render('FrameworkBundle...', Array)
#26 /var/www/html/vendor/symfony/form/Extension/Templating/TemplatingRendererEngine.php(41): Mautic\\CoreBundle\\Templating\\Engine\\PhpEngine->render('FrameworkBundle...', Array)
#27 /var/www/html/vendor/symfony/form/FormRenderer.php(260): Symfony\\Component\\Form\\Extension\\Templating\\TemplatingRendererEngine->renderBlock(Object(Symfony\\Component\\Form\\FormView), 'FrameworkBundle...', 'form_widget', Array)
#28 /var/www/html/vendor/symfony/framework-bundle/Templating/Helper/FormHelper.php(137): Symfony\\Component\\Form\\FormRenderer->searchAndRenderBlock(Object(Symfony\\Component\\Form\\FormView), 'widget', Array)
#29 /var/www/html/app/bundles/UserBundle/Views/User/form.html.php(87): Symfony\\Bundle\\FrameworkBundle\\Templating\\Helper\\FormHelper->widget(Object(Symfony\\Component\\Form\\FormView), Array)
#30 /var/www/html/app/bundles/CoreBundle/Templating/Engine/PhpEngine.php(147): require('/var/www/html/a...')
#31 /var/www/html/vendor/symfony/templating/PhpEngine.php(76): Mautic\\CoreBundle\\Templating\\Engine\\PhpEngine->evaluate(NULL, Array)
#32 /var/www/html/app/bundles/CoreBundle/Templating/Engine/PhpEngine.php(119): Symfony\\Component\\Templating\\PhpEngine->render('MauticUserBundl...', Array)
#33 /var/www/html/vendor/symfony/templating/DelegatingEngine.php(41): Mautic\\CoreBundle\\Templating\\Engine\\PhpEngine->render('MauticUserBundl...', Array)
#34 /var/www/html/vendor/symfony/framework-bundle/Controller/ControllerTrait.php(213): Symfony\\Component\\Templating\\DelegatingEngine->render('MauticUserBundl...', Array)
#35 /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php(364): Symfony\\Bundle\\FrameworkBundle\\Controller\\AbstractController->renderView('MauticUserBundl...', Array)
#36 /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php(197): Mautic\\CoreBundle\\Controller\\CommonController->ajaxAction(Array)
#37 /var/www/html/app/bundles/UserBundle/Controller/UserController.php(198): Mautic\\CoreBundle\\Controller\\CommonController->delegateView(Array)
#38 /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php(442): Mautic\\UserBundle\\Controller\\UserController->newAction(0, '')
#39 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(159): Mautic\\CoreBundle\\Controller\\CommonController->executeAction('new', 0, 0, '')
#40 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(81): Symfony\\Component\\HttpKernel\\HttpKernel->handleRaw(Object(Symfony\\Component\\HttpFoundation\\Request), 1)
#41 /var/www/html/vendor/symfony/http-kernel/Kernel.php(201): Symfony\\Component\\HttpKernel\\HttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#42 /var/www/html/app/AppKernel.php(105): Symfony\\Component\\HttpKernel\\Kernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#43 /var/www/html/app/middlewares/CORSMiddleware.php(82): AppKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#44 /var/www/html/app/middlewares/CatchExceptionMiddleware.php(34): Mautic\\Middleware\\CORSMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#45 /var/www/html/app/middlewares/Dev/IpRestrictMiddleware.php(52): Mautic\\Middleware\\CatchExceptionMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#46 /var/www/html/app/middlewares/VersionCheckMiddleware.php(58): Mautic\\Middleware\\Dev\\IpRestrictMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#47 /var/www/html/app/middlewares/TrustMiddleware.php(42): Mautic\\Middleware\\VersionCheckMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#48 /var/www/html/vendor/stack/builder/src/Stack/StackedHttpKernel.php(23): Mautic\\Middleware\\TrustMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#49 /var/www/html/vendor/stack/run/src/Stack/run.php(13): Stack\\StackedHttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request))
#50 /var/www/html/index_dev.php(25): Stack\
un(Object(Stack\\StackedHttpKernel))
#51 {main}
"} {"hostname":"mautic-web","pid":3516}

```

3. When removing a channel from a contact manually (via preferences on the contact) the following error occurs:

```
[2022-10-26 13:56:14] mautic.CRITICAL: Uncaught PHP Exception Error: "Call to a member function hasModel() on null" at /var/www/html/app/bundles/CoreBundle/EventListener/ChannelTrait.php line 28 {"exception":"[object] (Error(code: 0): Call to a member function hasModel() on null at /var/www/html/app/bundles/CoreBundle/EventListener/ChannelTrait.php:28)
[stacktrace]
#0 /var/www/html/app/bundles/LeadBundle/EventListener/LeadSubscriber.php(569): Mautic\\LeadBundle\\EventListener\\LeadSubscriber->getChannelModel('email')
#1 /var/www/html/app/bundles/LeadBundle/EventListener/LeadSubscriber.php(361): Mautic\\LeadBundle\\EventListener\\LeadSubscriber->addTimelineDoNotContactEntries(Object(Mautic\\LeadBundle\\Event\\LeadTimelineEvent), 'lead.donotconta...', 'Do not contact')
#2 /var/www/html/vendor/symfony/event-dispatcher/Debug/WrappedListener.php(126): Mautic\\LeadBundle\\EventListener\\LeadSubscriber->onTimelineGenerate(Object(Mautic\\LeadBundle\\Event\\LeadTimelineEvent), 'mautic.lead_tim...', Object(Symfony\\Component\\HttpKernel\\Debug\\TraceableEventDispatcher))
#3 /var/www/html/vendor/symfony/event-dispatcher/EventDispatcher.php(251): Symfony\\Component\\EventDispatcher\\Debug\\WrappedListener->__invoke(Object(Mautic\\LeadBundle\\Event\\LeadTimelineEvent), 'mautic.lead_tim...', Object(Symfony\\Component\\HttpKernel\\Debug\\TraceableEventDispatcher))
#4 /var/www/html/vendor/symfony/event-dispatcher/EventDispatcher.php(73): Symfony\\Component\\EventDispatcher\\EventDispatcher->callListeners(Array, 'mautic.lead_tim...', Object(Mautic\\LeadBundle\\Event\\LeadTimelineEvent))
#5 /var/www/html/vendor/symfony/event-dispatcher/Debug/TraceableEventDispatcher.php(168): Symfony\\Component\\EventDispatcher\\EventDispatcher->dispatch(Object(Mautic\\LeadBundle\\Event\\LeadTimelineEvent), 'mautic.lead_tim...')
#6 /var/www/html/app/bundles/LeadBundle/Model/LeadModel.php(2083): Symfony\\Component\\EventDispatcher\\Debug\\TraceableEventDispatcher->dispatch(Object(Mautic\\LeadBundle\\Event\\LeadTimelineEvent), 'mautic.lead_tim...')
#7 /var/www/html/app/bundles/LeadBundle/Controller/LeadDetailsTrait.php(305): Mautic\\LeadBundle\\Model\\LeadModel->getEngagements(Object(Mautic\\LeadBundle\\Entity\\Lead), Array, Array, 1, 25)
#8 /var/www/html/app/bundles/LeadBundle/Controller/LeadController.php(381): Mautic\\LeadBundle\\Controller\\LeadController->getEngagements(Object(Mautic\\LeadBundle\\Entity\\Lead))
#9 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(159): Mautic\\LeadBundle\\Controller\\LeadController->viewAction('999')
#10 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(81): Symfony\\Component\\HttpKernel\\HttpKernel->handleRaw(Object(Symfony\\Component\\HttpFoundation\\Request), 2)
#11 /var/www/html/vendor/symfony/framework-bundle/Controller/ControllerTrait.php(96): Symfony\\Component\\HttpKernel\\HttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 2)
#12 /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php(355): Symfony\\Bundle\\FrameworkBundle\\Controller\\AbstractController->forward('Mautic\\\\LeadBund...', Array, Array)
#13 /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php(285): Mautic\\CoreBundle\\Controller\\CommonController->ajaxAction(Array)
#14 /var/www/html/app/bundles/LeadBundle/Controller/LeadController.php(1004): Mautic\\CoreBundle\\Controller\\CommonController->postActionRedirect(Array)
#15 /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php(442): Mautic\\LeadBundle\\Controller\\LeadController->contactFrequencyAction('999', '')
#16 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(159): Mautic\\CoreBundle\\Controller\\CommonController->executeAction('contactFrequenc...', '999', 0, '')
#17 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(81): Symfony\\Component\\HttpKernel\\HttpKernel->handleRaw(Object(Symfony\\Component\\HttpFoundation\\Request), 1)
#18 /var/www/html/vendor/symfony/http-kernel/Kernel.php(201): Symfony\\Component\\HttpKernel\\HttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#19 /var/www/html/app/AppKernel.php(105): Symfony\\Component\\HttpKernel\\Kernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#20 /var/www/html/app/middlewares/CORSMiddleware.php(82): AppKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#21 /var/www/html/app/middlewares/CatchExceptionMiddleware.php(34): Mautic\\Middleware\\CORSMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#22 /var/www/html/app/middlewares/Dev/IpRestrictMiddleware.php(52): Mautic\\Middleware\\CatchExceptionMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#23 /var/www/html/app/middlewares/VersionCheckMiddleware.php(58): Mautic\\Middleware\\Dev\\IpRestrictMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#24 /var/www/html/app/middlewares/TrustMiddleware.php(42): Mautic\\Middleware\\VersionCheckMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#25 /var/www/html/vendor/stack/builder/src/Stack/StackedHttpKernel.php(23): Mautic\\Middleware\\TrustMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#26 /var/www/html/vendor/stack/run/src/Stack/run.php(13): Stack\\StackedHttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request))
#27 /var/www/html/index_dev.php(25): Stack\
un(Object(Stack\\StackedHttpKernel))
#28 {main}
"} {"hostname":"mautic-web","pid":3462}
```

4. Segment emails are not being sent - maybe need to pull in the latest changes to get the fix we merged recently?

File: app/AppKernel.php [hunk 1]
```diff
@@ -219,6 +220,12 @@ public function registerBundles(): array
         return $bundles;
     }
 
+    protected function build(ContainerBuilder $container): void
+    {
+        $container->registerForAutoconfiguration(\Mautic\CoreBundle\Model\MauticModelInterface::class)
```
Comment (biozshock) [inline app/AppKernel.php:225]: This must be added to a documentation.

**Comment by biozshock** [comment]
> @biozshock I tried the trait you suggested but it creates a loop:
> 
> ```
>   [Symfony\Component\DependencyInjection\Exception\ServiceCircularReferenceException]                                                                       
>   Circular reference detected for service "LightSaml\Store\Credential\CompositeCredentialStore", path: "LightSaml\Store\Credential\CompositeCredentialStore -> LightSaml\Store\Credential\CompositeCredentialStore". 
> ```

Hm, are you sure you added this to the DI extension class in the `load` method? Before compiling the container wouldn't know anything about the Ligtsaml classes, and the trait will process only services added in the loading process.

BTW: Actually after refactoring the aliasing should be done only in the `app/bundles/CoreBundle/DependencyInjection/Compiler/ServicePass.php`. I didn't look deeper into what the pass actually does, but it seems it already creates class aliases if the `id` of the service is not the classFQCN here https://github.com/mautic/mautic/pull/11587/files#diff-c970d70821380696974e49f7cef5ee199e8124dda95d21e2b2e5fdd25e29b5c7R117 ?

**Comment by escopecz** [comment]
@RCheesley Thanks for you tests! I fixed the issues. The one with the builder took me quite a while to figure out. The problem was that the AssetsHelper was instantiated twice. The first instance got the builder JS but the other instance was asked to print it to the HTML.

**Comment by escopecz** [comment]
Rebased to resolve conflicts.

**Review by mabumusa1 — DISMISSED** [review]
I used gitpod to test the controllers, and every controller I tested worked without issues. Mautic loaded without.

I think this PR is good to go

**Review by volha-pivavarchyk — DISMISSED** [review]
Works well for me locally and on Gitpod

**Review by biozshock — DISMISSED** [review]
A leap forward!
Created couple tickets instead of commenting on the code.

File: app/AppKernel.php (hunk 1 above)
Comment (escopecz) [reply app/AppKernel.php:225]: Nothing has changed for the model classes themselves. They are "Mautic models" when they extend `AbstractCommonModel` and `AbstractCommonModel implements MauticModelInterface` so all existing and new "Mautic models" will be autoconfigured. What would you suggest to add to the docs?

Comment (biozshock) [reply app/AppKernel.php:225]: Oh, i have overlooked that all models should extend `AbstractCommonModel` to actually be models in Mautic.

**Pull request merged**